File: apps/challenge-service/src/infrastructure/nats-event-publisher.ts
import type { ChallengeDto, DomainEvent } from "@sparkflow/contracts";
import { connect, JSONCodec, type JetStreamClient } from "nats";
import type { EventPublisher } from "../application/ports.ts";

const jsonCodec = JSONCodec<DomainEvent>();

export type NatsEventPublisher = EventPublisher & {
  readonly close: () => Promise<void>;
};

export const createNatsEventPublisher = async (input: {
  readonly natsUrl: string;
}): Promise<NatsEventPublisher> => {
  const connection = await connect({ servers: input.natsUrl });
  const jetStreamManager = await connection.jetstreamManager();

  try {
    await jetStreamManager.streams.add({
      name: "SPARKFLOW_EVENTS",
      subjects: ["challenge.*", "submission.*", "evaluation.*", "notification.*"],
    });
  } catch {
    await jetStreamManager.streams.info("SPARKFLOW_EVENTS");
  }

  const jetStream: JetStreamClient = connection.jetstream();

  return {
    close: async () => {
      await connection.drain();
    },
    publish: async ({ event }: { readonly event: DomainEvent<ChallengeDto> }) => {
      await jetStream.publish(event.eventName, jsonCodec.encode(event));
    },
  };
};

File: apps/evaluation-service/src/infrastructure/nats_event_publisher.py
import nats
from nats.aio.client import Client

from infrastructure.contracts import DomainEvent


class NatsEventPublisher:
    def __init__(self, *, connection: Client, jetstream: nats.js.JetStreamContext) -> None:
        self._connection = connection
        self._jetstream = jetstream

    async def close(self) -> None:
        await self._connection.drain()

    async def publish(self, *, event: DomainEvent) -> None:
        payload = event.model_dump_json(by_alias=True).encode("utf-8")
        await self._jetstream.publish(event.eventName, payload)


async def create_nats_event_publisher(*, nats_url: str) -> NatsEventPublisher:
    connection = await nats.connect(nats_url)
    jetstream = connection.jetstream()

    try:
        await jetstream.add_stream(
            name="SPARKFLOW_EVENTS",
            subjects=["challenge.*", "submission.*", "evaluation.*", "notification.*"],
        )
    except Exception:
        await jetstream.stream_info("SPARKFLOW_EVENTS")

    return NatsEventPublisher(connection=connection, jetstream=jetstream)
